test/Microsoft.AspNetCore.Razor.Language.Test/Legacy/MarkupElementRewriterTest.cs

```diff
@@ -0,0 +1,123 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Text;
+using Microsoft.AspNetCore.Razor.Language.Syntax;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Razor.Language.Legacy
+{
+    public class MarkupElementRewriterTest : CsHtmlMarkupParserTestBase
+    {
+        public MarkupElementRewriterTest()
+        {
+            UseNewSyntaxTree = true;
+        }
+
+        [Fact]
+        public void Rewrites_ValidTags()
+        {
+            // Arrange
+            var content = @"
+<div>Foo</div>
+<p>Bar</p>
+";
+
+            // Act & Assert
+            RewriterTest(content);
+        }
+
+        [Fact]
+        public void Rewrites_ValidNestedTags()
+        {
+            // Arrange
+            var content = @"
+<div>
+    Foo
+    <p>Bar</p>
+    Baz
+</div>";
+
+            // Act & Assert
+            RewriterTest(content);
+        }
+
+        [Fact]
+        public void Rewrites_ValidNestedTagsMixedWithCode()
+        {
+            // Arrange
+            var content = @"
+<div>
+    Foo
+    <p>@Bar</p>
+    @{ var x = Bar; }
+</div>
+";
+
+            // Act & Assert
+            RewriterTest(content);
+        }
+
+        [Fact]
+        public void Rewrites_EndTagsWithMissingStartTags()
+        {
+            // Arrange
+            var content = @"
+Foo</div>
+";
+
+            // Act & Assert
+            RewriterTest(content);
+        }
+
+        [Fact]
+        public void Rewrites_StartTagsWithMissingEndTags()
+        {
+            // Arrange
+            var content = @"
+<div>
+    Foo
+    <p>
+        Bar
+        </strong>
+";
+
+            // Act & Assert
+            RewriterTest(content);
+        }
+
+        [Fact]
+        public void Rewrites_VoidTagsAndSelfClosingTags()
+        {
+            // Arrange
+            var content = @"
+<br>Foo<custom />
+";
+
+            // Act & Assert
+            RewriterTest(content);
+        }
+
+        [Fact]
+        public void Rewrites_MalformedTags_RecoversSuccessfully()
+        {
+            // Arrange
+            var content = @"
+<div>content</span>footer</div>
+";
+
+            // Act & Assert
+            RewriterTest(content);
+        }
+
+        private void RewriterTest(string input)
+        {
+            var syntaxTree = ParseDocument(input, designTime: false);
+            var rewritten = MarkupElementRewriter.Rewrite(syntaxTree);
+            BaselineTest(rewritten);
+
+            var unrewritten = MarkupElementRewriter.RemoveMarkupElement(rewritten);
+            Assert.Equal(syntaxTree.Root.SerializedValue, unrewritten.Root.SerializedValue);
```

Comment:
Each test will also run the removal phase and make sure the output matches the original syntax tree. It didn't make sense to add separate tests for the removal phase.

Reply:
Love it.